In [1]:
#Importing all Librabries
import pandas as pd
import numpy as np
import matplotlib.pyplot as ply
import seaborn as sns
%matplotlib inline

In [2]:
df =pd.read_csv('https://archive.ics.uci.edu/ml/machine-learning-databases/00374/energydata_complete.csv')

In [3]:
df.head()

date  Appliances  lights     T1       RH_1    T2       RH_2  \
0  2016-01-11 17:00:00          60      30  19.89  47.596667  19.2  44.790000   
1  2016-01-11 17:10:00          60      30  19.89  46.693333  19.2  44.722500   
2  2016-01-11 17:20:00          50      30  19.89  46.300000  19.2  44.626667   
3  2016-01-11 17:30:00          50      40  19.89  46.066667  19.2  44.590000   
4  2016-01-11 17:40:00          60      40  19.89  46.333333  19.2  44.530000   

      T3       RH_3         T4  ...         T9   RH_9     T_out  Press_mm_hg  \
0  19.79  44.730000  19.000000  ...  17.033333  45.53  6.600000        733.5   
1  19.79  44.790000  19.000000  ...  17.066667  45.56  6.483333        733.6   
2  19.79  44.933333  18.926667  ...  17.000000  45.50  6.366667        733.7   
3  19.79  45.000000  18.890000  ...  17.000000  45.40  6.250000        733.8   
4  19.79  45.000000  18.890000  ...  17.000000  45.40  6.133333        733.9   

   RH_out  Windspeed  Visibility  Tdewpoint        rv1        rv2  
0    92.0   7.000000   63.000000        5.3  13.275433  13.275433  
1    92.0   6.666667   59.166667        5.2  18.606195  18.606195  
2    92.0   6.333333   55.333333        5.1  28.642668  28.642668  
3    92.0   6.000000   51.500000        5.0  45.410389  45.410389  
4    92.0   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 29 columns]

# Obtaining all necessaries information from the dataset

In [4]:
df.describe()

Appliances        lights            T1          RH_1            T2  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      97.694958      3.801875     21.686571     40.259739     20.341219   
std      102.524891      7.935988      1.606066      3.979299      2.192974   
min       10.000000      0.000000     16.790000     27.023333     16.100000   
25%       50.000000      0.000000     20.760000     37.333333     18.790000   
50%       60.000000      0.000000     21.600000     39.656667     20.000000   
75%      100.000000      0.000000     22.600000     43.066667     21.500000   
max     1080.000000     70.000000     26.260000     63.360000     29.856667   

               RH_2            T3          RH_3            T4          RH_4  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      40.420420     22.267611     39.242500     20.855335     39.026904   
std        4.069813      2.006111      3.254576      2.042884      4.341321   
min       20.463333     17.200000     28.766667     15.100000     27.660000   
25%       37.900000     20.790000     36.900000     19.530000     35.530000   
50%       40.500000     22.100000     38.530000     20.666667     38.400000   
75%       43.260000     23.290000     41.760000     22.100000     42.156667   
max       56.026667     29.236000     50.163333     26.200000     51.090000   

       ...            T9          RH_9         T_out   Press_mm_hg  \
count  ...  19735.000000  19735.000000  19735.000000  19735.000000   
mean   ...     19.485828     41.552401      7.411665    755.522602   
std    ...      2.014712      4.151497      5.317409      7.399441   
min    ...     14.890000     29.166667     -5.000000    729.300000   
25%    ...     18.000000     38.500000      3.666667    750.933333   
50%    ...     19.390000     40.900000      6.916667    756.100000   
75%    ...     20.600000     44.338095     10.408333    760.933333   
max    ...     24.500000     53.326667     26.100000    772.300000   

             RH_out     Windspeed    Visibility     Tdewpoint           rv1  \
count  19735.000000  19735.000000  19735.000000  19735.000000  19735.000000   
mean      79.750418      4.039752     38.330834      3.760707     24.988033   
std       14.901088      2.451221     11.794719      4.194648     14.496634   
min       24.000000      0.000000      1.000000     -6.600000      0.005322   
25%       70.333333      2.000000     29.000000      0.900000     12.497889   
50%       83.666667      3.666667     40.000000      3.433333     24.897653   
75%       91.666667      5.500000     40.000000      6.566667     37.583769   
max      100.000000     14.000000     66.000000     15.500000     49.996530   

                rv2  
count  19735.000000  
mean      24.988033  
std       14.496634  
min        0.005322  
25%       12.497889  
50%       24.897653  
75%       37.583769  
max       49.996530  

[8 rows x 28 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 19735 entries, 0 to 19734
Data columns (total 29 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   date         19735 non-null  object 
 1   Appliances   19735 non-null  int64  
 2   lights       19735 non-null  int64  
 3   T1           19735 non-null  float64
 4   RH_1         19735 non-null  float64
 5   T2           19735 non-null  float64
 6   RH_2         19735 non-null  float64
 7   T3           19735 non-null  float64
 8   RH_3         19735 non-null  float64
 9   T4           19735 non-null  float64
 10  RH_4         19735 non-null  float64
 11  T5           19735 non-null  float64
 12  RH_5         19735 non-null  float64
 13  T6           19735 non-null  float64
 14  RH_6         19735 non-null  float64
 15  T7           19735 non-null  float64
 16  RH_7         19735 non-null  float64
 17  T8           19735 non-null  float64
 18  RH_8         19735 non-null  float64
 19  T9  

# QUESTION 12

In [6]:
from sklearn.model_selection import train_test_split

In [7]:
X = df['T2'].values

In [8]:
y = df['T6'].values

In [9]:
X_train,X_test,y_train,y_test =train_test_split(X,y,test_size=0.3,random_state=42)

In [10]:
from sklearn.linear_model import LinearRegression

In [11]:
lm = LinearRegression()

In [12]:
# line of code is done to convert the Variables of both the target(y) and the feature(x) to 
# a 2D array from its original 1D array
# This is done to avoid error message


X_train = X_train.reshape(-1,1)
y_train = y_train.reshape(-1,1)
X_test = X_test.reshape(-1,1)
y_test = y_test.reshape(-1,1)

In [13]:
lm.fit(X_train,y_train)

LinearRegression()

In [14]:
predictions = lm.predict(X_test)

In [15]:
predictions

array([[ 2.15578912],
       [10.01116055],
       [ 1.87391554],
       ...,
       [ 4.24758774],
       [ 8.69822311],
       [ 4.9893603 ]])

In [16]:
#importing metrics to see the outcome of trained Data

In [17]:
from sklearn import metrics

In [18]:
#Calculating R Squared

In [19]:
r2 = metrics.r2_score(y_test,predictions)

In [20]:
r2

0.6427789637197382

In [21]:
#Rounding off to 2DP

In [22]:
round(r2,2)

0.64

#The answer is 0.64

# Question 13

In [23]:
#First thing to do is to normalize our dataset

In [24]:
from sklearn.preprocessing import MinMaxScaler

In [25]:
scaler = MinMaxScaler()

In [26]:
#According to instruction, we are asked to drop colums['date','lights']

In [27]:
df.drop(columns=['date','lights'], inplace=True)

In [28]:
df.head()

Appliances     T1       RH_1    T2       RH_2     T3       RH_3         T4  \
0          60  19.89  47.596667  19.2  44.790000  19.79  44.730000  19.000000   
1          60  19.89  46.693333  19.2  44.722500  19.79  44.790000  19.000000   
2          50  19.89  46.300000  19.2  44.626667  19.79  44.933333  18.926667   
3          50  19.89  46.066667  19.2  44.590000  19.79  45.000000  18.890000   
4          60  19.89  46.333333  19.2  44.530000  19.79  45.000000  18.890000   

        RH_4         T5  ...         T9   RH_9     T_out  Press_mm_hg  RH_out  \
0  45.566667  17.166667  ...  17.033333  45.53  6.600000        733.5    92.0   
1  45.992500  17.166667  ...  17.066667  45.56  6.483333        733.6    92.0   
2  45.890000  17.166667  ...  17.000000  45.50  6.366667        733.7    92.0   
3  45.723333  17.166667  ...  17.000000  45.40  6.250000        733.8    92.0   
4  45.530000  17.200000  ...  17.000000  45.40  6.133333        733.9    92.0   

   Windspeed  Visibility  Tdewpoint        rv1        rv2  
0   7.000000   63.000000        5.3  13.275433  13.275433  
1   6.666667   59.166667        5.2  18.606195  18.606195  
2   6.333333   55.333333        5.1  28.642668  28.642668  
3   6.000000   51.500000        5.0  45.410389  45.410389  
4   5.666667   47.666667        4.9  10.084097  10.084097  

[5 rows x 27 columns]

In [29]:
normalized_df = pd.DataFrame(data = scaler.fit_transform(df), columns=df.columns)

In [30]:
normalized_df.head()

Appliances       T1      RH_1        T2      RH_2        T3      RH_3  \
0    0.046729  0.32735  0.566187  0.225345  0.684038  0.215188  0.746066   
1    0.046729  0.32735  0.541326  0.225345  0.682140  0.215188  0.748871   
2    0.037383  0.32735  0.530502  0.225345  0.679445  0.215188  0.755569   
3    0.037383  0.32735  0.524080  0.225345  0.678414  0.215188  0.758685   
4    0.046729  0.32735  0.531419  0.225345  0.676727  0.215188  0.758685   

         T4      RH_4        T5  ...        T9      RH_9     T_out  \
0  0.351351  0.764262  0.175506  ...  0.223032  0.677290  0.372990   
1  0.351351  0.782437  0.175506  ...  0.226500  0.678532  0.369239   
2  0.344745  0.778062  0.175506  ...  0.219563  0.676049  0.365488   
3  0.341441  0.770949  0.175506  ...  0.219563  0.671909  0.361736   
4  0.341441  0.762697  0.178691  ...  0.219563  0.671909  0.357985   

   Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1       rv2  
0     0.097674  0.894737   0.500000    0.953846   0.538462  0.265449  0.265449  
1     0.100000  0.894737   0.476190    0.894872   0.533937  0.372083  0.372083  
2     0.102326  0.894737   0.452381    0.835897   0.529412  0.572848  0.572848  
3     0.104651  0.894737   0.428571    0.776923   0.524887  0.908261  0.908261  
4     0.106977  0.894737   0.404762    0.717949   0.520362  0.201611  0.201611  

[5 rows x 27 columns]

In [31]:
# As instructed 'Appliances' is our target Variable

In [32]:
normalized_df.columns

Index(['Appliances', 'T1', 'RH_1', 'T2', 'RH_2', 'T3', 'RH_3', 'T4', 'RH_4',
       'T5', 'RH_5', 'T6', 'RH_6', 'T7', 'RH_7', 'T8', 'RH_8', 'T9', 'RH_9',
       'T_out', 'Press_mm_hg', 'RH_out', 'Windspeed', 'Visibility',
       'Tdewpoint', 'rv1', 'rv2'],
      dtype='object')

In [33]:
#copying all in columns aside Appliances which is our target 
features_df =normalized_df[['T1', 'RH_1', 'T2', 'RH_2', 'T3', 'RH_3', 'T4', 'RH_4', 'T5', 'RH_5', 'T6', 'RH_6', 'T7', 'RH_7', 'T8', 'RH_8', 'T9', 'RH_9', 'T_out', 'Press_mm_hg', 'RH_out', 'Windspeed', 'Visibility','Tdewpoint', 'rv1', 'rv2']]

In [34]:
features_df

T1      RH_1        T2      RH_2        T3      RH_3        T4  \
0      0.327350  0.566187  0.225345  0.684038  0.215188  0.746066  0.351351   
1      0.327350  0.541326  0.225345  0.682140  0.215188  0.748871  0.351351   
2      0.327350  0.530502  0.225345  0.679445  0.215188  0.755569  0.344745   
3      0.327350  0.524080  0.225345  0.678414  0.215188  0.758685  0.341441   
4      0.327350  0.531419  0.225345  0.676727  0.215188  0.758685  0.341441   
...         ...       ...       ...       ...       ...       ...       ...   
19730  0.926786  0.537657  0.711655  0.606309  0.830841  0.579374  0.864865   
19731  0.919747  0.536006  0.701769  0.607836  0.825302  0.582178  0.864865   
19732  0.919747  0.538666  0.692651  0.627198  0.818378  0.603988  0.864865   
19733  0.919747  0.549491  0.677054  0.634717  0.805085  0.585294  0.864865   
19734  0.919747  0.538758  0.666171  0.632902  0.799546  0.579062  0.864865   

           RH_4        T5      RH_5  ...        T9      RH_9     T_out  \
0      0.764262  0.175506  0.381691  ...  0.223032  0.677290  0.372990   
1      0.782437  0.175506  0.381691  ...  0.226500  0.678532  0.369239   
2      0.778062  0.175506  0.380037  ...  0.219563  0.676049  0.365488   
3      0.770949  0.175506  0.380037  ...  0.219563  0.671909  0.361736   
4      0.762697  0.178691  0.380037  ...  0.219563  0.671909  0.357985   
...         ...       ...       ...  ...       ...       ...       ...   
19730  0.765258  0.752031  0.339590  ...  0.864724  0.729443  0.891747   
19731  0.765258  0.754897  0.338487  ...  0.864724  0.729443  0.887460   
19732  0.771233  0.754897  0.337585  ...  0.864724  0.729443  0.883173   
19733  0.773794  0.752031  0.336583  ...  0.864724  0.730581  0.878885   
19734  0.781192  0.752031  0.336583  ...  0.864724  0.731719  0.874598   

       Press_mm_hg    RH_out  Windspeed  Visibility  Tdewpoint       rv1  \
0         0.097674  0.894737   0.500000    0.953846   0.538462  0.265449   
1         0.100000  0.894737   0.476190    0.894872   0.533937  0.372083   
2         0.102326  0.894737   0.452381    0.835897   0.529412  0.572848   
3         0.104651  0.894737   0.428571    0.776923   0.524887  0.908261   
4         0.106977  0.894737   0.404762    0.717949   0.520362  0.201611   
...            ...       ...        ...         ...        ...       ...   
19730     0.602326  0.416667   0.238095    0.348718   0.901961  0.861981   
19731     0.602326  0.421053   0.250000    0.361538   0.900452  0.985726   
19732     0.602326  0.425439   0.261905    0.374359   0.898944  0.583979   
19733     0.602326  0.429825   0.273810    0.387179   0.897436  0.126371   
19734     0.602326  0.434211   0.285714    0.400000   0.895928  0.682391   

            rv2  
0      0.265449  
1      0.372083  
2      0.572848  
3      0.908261  
4      0.201611  
...         ...  
19730  0.861981  
19731  0.985726  
19732  0.583979  
19733  0.126371  
19734  0.682391  

[19735 rows x 26 columns]

In [35]:
target_df =normalized_df['Appliances']

In [36]:
from sklearn.model_selection import train_test_split

In [37]:
X_train,X_test,y_train,y_test = train_test_split(features_df,target_df, test_size=0.3, random_state=42)

In [38]:
#calling linear Regresion
from sklearn.linear_model import LinearRegression

In [39]:
lm2= LinearRegression()

In [40]:
lm2.fit(X_train,y_train)

LinearRegression()

In [41]:
pdt = lm2.predict(X_test)

In [42]:
pdt 

array([0.03322207, 0.24411599, 0.03400024, ..., 0.06844707, 0.10032325,
       0.05722198])

In [43]:
#Calculating the Mean Absolute Error (MAE)

In [44]:
from sklearn import metrics

In [45]:
MAE = metrics.mean_absolute_error(y_test,pdt)

In [46]:
MAE

0.05013362397742955

In [47]:
#Rounding off to 2DP
MAE_2DP =round(MAE,2)

In [48]:
MAE_2DP

0.05

# QUESTION 14

#What is the Residual Sum of Squares (in two decimal places)?

In [49]:
# RSS values
RSS = np.sum(np.square(y_test - pdt))
RSS
round(RSS,2)

45.35

# QUESTION 15

In [50]:
#calculating the the Root Mean Squared Error (in three decimal places)
RMS= np.sqrt(metrics.mean_squared_error(y_test,pdt))

In [51]:
RMS

0.0875144494766171

In [52]:
#Rounding it off to 3DP

In [53]:
round(RMS,3)

0.088

# QUESTION 16

In [55]:
#We are asked to find the coefficient of Determination in 2DP

In [57]:
# R Squared values also called Coefficient of Determination
COD = metrics.r2_score(y_test,pdt)


In [58]:
COD

0.14890246319303535

In [59]:
#Rounding it off to 2DP

In [60]:
round(COD,2)

0.15

Coeficient of Determination is 0.15

# QUESTION 17

In [61]:
#We asked to obtain the feature weights from your linear model above.
#Which features have the lowest and highest weights respectively?

In [62]:
# comparing the effects of regularisation
def get_weights_df(model, feat, col_name):
  weights = pd.Series(model.coef_, feat.columns).sort_values()
  weights_df = pd.DataFrame(weights).reset_index()
  weights_df.columns = ['Features', col_name]
  weights_df[col_name].round(3)
  return weights_df

In [63]:
# Importing the Lasso regression

from sklearn.linear_model import Lasso
lasso_reg = Lasso(alpha=0.001)
lasso_reg.fit(X_train, y_train)

Lasso(alpha=0.001)

In [64]:
# Importing the Ridge regression

from sklearn.linear_model import Ridge
ridge_reg = Ridge(alpha=0.4)
ridge_reg.fit(X_train, y_train)

Ridge(alpha=0.4)

In [67]:
linear_model_weights = get_weights_df(lm2, X_train, 'Linear_Model_Weight')
ridge_weights_df = get_weights_df(ridge_reg, X_train, 'Ridge_Weight')
lasso_weights_df = get_weights_df(lasso_reg, X_train, 'Lasso_weight')

In [68]:
final_weights = pd.merge(linear_model_weights, ridge_weights_df, on='Features')
final_weights = pd.merge(final_weights, lasso_weights_df, on='Features')

In [76]:
final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0          RH_2            -0.456698     -0.411071     -0.000000
1         T_out            -0.321860     -0.262172      0.000000
2            T2            -0.236178     -0.201397      0.000000
3            T9            -0.189941     -0.188916     -0.000000
4          RH_8            -0.157595     -0.156830     -0.000110
5        RH_out            -0.077671     -0.054724     -0.049557
6          RH_7            -0.044614     -0.045977     -0.000000
7          RH_9            -0.039800     -0.041367     -0.000000
8            T5            -0.015657     -0.019853     -0.000000
9            T1            -0.003281     -0.018406      0.000000
10          rv1             0.000770      0.000748     -0.000000
11          rv2             0.000770      0.000748     -0.000000
12  Press_mm_hg             0.006839      0.006584     -0.000000
13           T7             0.010319      0.010098     -0.000000
14   Visibility             0.012307      0.012076      0.000000
15         RH_5             0.016006      0.016152      0.000000
16         RH_4             0.026386      0.024579      0.000000
17           T4             0.028981      0.027384     -0.000000
18    Windspeed             0.029183      0.030268      0.002912
19         RH_6             0.038049      0.035519     -0.000000
20         RH_3             0.096048      0.095135      0.000000
21           T8             0.101995      0.101028      0.000000
22    Tdewpoint             0.117758      0.083128      0.000000
23           T6             0.236425      0.217292      0.000000
24           T3             0.290627      0.288087      0.000000
25         RH_1             0.553547      0.519525      0.017880

In [77]:
#From the outcome

RH_2 has the lowest weights of -0.456698

RH_1 has the highest weights of 0.553547

# QUESTION 18

In [79]:
ridge_prediction = ridge_reg.predict(X_test)

In [80]:
ridge_prediction

array([0.03321872, 0.24043824, 0.03461337, ..., 0.06872351, 0.10025536,
       0.05851175])

In [81]:
# RMSE
RMSE = np.sqrt(metrics.mean_squared_error(y_test,ridge_prediction))


In [83]:
RMSE

0.08753385704628003

In [84]:
round(RMSE,3)

0.088

# QUESTION 19

In [89]:
final_weights

Features  Linear_Model_Weight  Ridge_Weight  Lasso_weight
0          RH_2            -0.456698     -0.411071     -0.000000
1         T_out            -0.321860     -0.262172      0.000000
2            T2            -0.236178     -0.201397      0.000000
3            T9            -0.189941     -0.188916     -0.000000
4          RH_8            -0.157595     -0.156830     -0.000110
5        RH_out            -0.077671     -0.054724     -0.049557
6          RH_7            -0.044614     -0.045977     -0.000000
7          RH_9            -0.039800     -0.041367     -0.000000
8            T5            -0.015657     -0.019853     -0.000000
9            T1            -0.003281     -0.018406      0.000000
10          rv1             0.000770      0.000748     -0.000000
11          rv2             0.000770      0.000748     -0.000000
12  Press_mm_hg             0.006839      0.006584     -0.000000
13           T7             0.010319      0.010098     -0.000000
14   Visibility             0.012307      0.012076      0.000000
15         RH_5             0.016006      0.016152      0.000000
16         RH_4             0.026386      0.024579      0.000000
17           T4             0.028981      0.027384     -0.000000
18    Windspeed             0.029183      0.030268      0.002912
19         RH_6             0.038049      0.035519     -0.000000
20         RH_3             0.096048      0.095135      0.000000
21           T8             0.101995      0.101028      0.000000
22    Tdewpoint             0.117758      0.083128      0.000000
23           T6             0.236425      0.217292      0.000000
24           T3             0.290627      0.288087      0.000000
25         RH_1             0.553547      0.519525      0.017880

# QUESTION 20

In [94]:
lasso_prediction = lasso_reg.predict(X_test)

In [95]:
lasso_prediction

array([0.07370267, 0.08143458, 0.07716072, ..., 0.07792848, 0.09034412,
       0.08359255])

In [96]:
RMSE = np.sqrt(metrics.mean_squared_error(y_test,lasso_pred))

In [97]:
RMSE

0.09358170467245137

In [98]:
#Rounding off to three Decimal places

In [100]:
NEW_RMSE =round(RMSE,3)

In [101]:
NEW_RMSE

0.094